framework/src/media/HttpInputDataSource.cpp

```diff
@@ -0,0 +1,250 @@
+	}
+
+	medvdbg("setAudioType %d\n", audioType);
+	setAudioType(audioType);
+
+	switch (audioType) {
+	case AUDIO_TYPE_MP3:
+	case AUDIO_TYPE_AAC: {
+		// wait for audio stream data
+		if (!mCondv.wait_for(lock, WAIT_DATA_TIMEOUT, [=]{return mIsDataReceived;})) {
+			medwdbg("download:: wait audio data timeout!\n");
+			mBufferWriter->setEndOfStream();
+			return false;
+		}
+
+		size_t templen = mBufferReader->sizeOfData();
+		assert(templen != 0);
+		unsigned char *tempbuf = new unsigned char[templen];
+		assert(tempbuf != nullptr);
+		size_t dlen = mBufferReader->dump(0, tempbuf, templen);
+		assert(templen == dlen);
+
+		unsigned int channel;
+		unsigned int sampleRate;
+		bool ret = utils::header_parsing(tempbuf, dlen, audioType, &channel, &sampleRate, NULL);
+		delete[] tempbuf;
+
+		if (!ret) {
+			meddbg("header parsing failed\n");
+			mBufferWriter->setEndOfStream();
+			return false;
+		}
+
+		medvdbg("setSampleRate:%u,setChannels:%u\n", sampleRate, channel);
+		setSampleRate(sampleRate);
+		setChannels(channel);
+		break;
+	}
+
+	default:
+		/* Don't set any decoder for unsupported formats */
+		break;
+	}
+
+	medvdbg("HttpInputDataSource::open! exit\n");
+	return true;
+}
+
+bool HttpInputDataSource::close()
+{
+	medvdbg("HttpInputDataSource::close enter\n");
+	mBufferWriter->setEndOfStream();
+	if (mThread.joinable()) {
+		mThread.join();
+	}
+	medvdbg("HttpInputDataSource::close exit!\n");
+	return true;
+}
+
+bool HttpInputDataSource::isPrepare()
+{
+	return (mStreamBuffer != nullptr && mHttpStream != nullptr);
+}
+
+ssize_t HttpInputDataSource::read(unsigned char *buf, size_t size)
+{
+	if (!isPrepare()) {
+		meddbg("[line:%d] Fail : HttpInputDataSource is not prepared\n", __LINE__);
+		return EOF;
+	}
+
+	if (buf == nullptr) {
+		meddbg("[line:%d] Fail : buf is nullptr\n", __LINE__);
+		return EOF;
+	}
+
+	size_t rlen = 0;
+	if (mBufferReader) {
+		rlen = mBufferReader->read(buf, size);
+	}
+	return rlen;
+}
+
+size_t HttpInputDataSource::HeaderCallback(char *data, size_t size, size_t nmemb, void *userp)
+{
+	auto source = static_cast<HttpInputDataSource *>(userp);
+	if (source->mBufferReader->isEndOfStream()) {
+		medwdbg("end-of-stream:true\n");
+		return 0;
+	}
+
+	size_t totalsize = size * nmemb;
+	std::string header(data, totalsize);
+	medvdbg("%s\n", header.c_str());
+	if (header.find(TAG_CONTENT_TYPE) != std::string::npos) {
+		source->mContentType = header.substr(TAG_CONTENT_TYPE.length());
+		if (!source->mIsHeaderReceived) {
+			std::lock_guard<std::mutex> lock(source->mMutex);
+			source->mIsHeaderReceived = true;
+			source->mCondv.notify_one();
+		}
+	}
+
+	return totalsize;
+}
+
+size_t HttpInputDataSource::WriteCallback(char *data, size_t size, size_t nmemb, void *userp)
+{
+	auto source = static_cast<HttpInputDataSource *>(userp);
+	size_t totalsize = size * nmemb;
+	size_t wlen = source->mBufferWriter->write((unsigned char *)data, totalsize);
+
```

Comment:
This is async function..
You should check the StreamBuffer if it is full or not
if full, than you need to wait(lock) until avail.

Reply:
Stream buffer writer already has such mechanism, it  will wait until all data is written.
```c
size_t StreamBufferWriter::write(unsigned char *buf, size_t size, bool sync = true);
```